### PR TITLE
Tenant scoping apis

### PR DESCRIPTION
**Pull request by bspates**
### Ticket #72 
Also fixes: 
* #129
* #61  

### Description
Re-scopes access on api endpoints from agency hierarchy to the rules specified in #72 

### Screenshots / Demo Video
Did bug bash with Mindy

### Testing
- Updated existing api/db unit tests to match new logic
- Manually tested UI use cases

### Checklist
- [x] Provided ticket and description
- [x] Your PR title contains the ISSUE ticket number e.g "86: ..."
- [x] Provided screenshots/demo
- [x] Provided testing information
- [x] Provided adequate test coverage for all new code
- [x] Added PR reviewers
- [x] Ensure at least 1 review before merging

### CODE REVIEW

File: packages/server/src/routes/agencies.js
Comment (mattbroussard) [inline packages/server/src/routes/agencies.js:24]: I'm confused about this endpoint:
1. It's not called anywhere? What is its intended use?
2. I thought based on the changes in access-helpers that we are now allowing impersonation of any agency in the tenant -- but this still uses `getAgencyTree` to return only subagencies of a user's agency.
3. What is the difference between `asAgency` and `organizationId` (aka `selectedAgency`)?

File: packages/server/src/routes/agencies.js
Comment (mattbroussard) [inline packages/server/src/routes/agencies.js:40]: nit: rename the path parameter to `:agencyId` to make its downstream use clearer. Same for the other routes in this file that have a `:agency` param.

File: packages/server/src/routes/agencies.js
Comment (mattbroussard) [inline packages/server/src/routes/agencies.js:118]: nit: send a 403 instead of throwing, which will send a 500.

File: packages/server/src/lib/access-helpers.js
Comment (mattbroussard) [inline packages/server/src/lib/access-helpers.js:10]: Naming of the parameter here and the way it's passed through to `inTenant` doesn't make it clear that it can be a single ID or a list -- as noted elsewhere, would prefer to just make it always a list and name it `agencyIds`.

Also: I'd consider renaming this function to be a bit more explicit (e.g. `isUserAuthorizedForAgency`) since it's now used in many more places outside this file where the context is less obvious about what authorization means.

File: packages/server/src/routes/tenants.js
Comment (mattbroussard) [inline packages/server/src/routes/tenants.js:9]: Note: will merge conflict with my changes in #189

File: packages/server/src/routes/users.js
Comment (mattbroussard) [inline packages/server/src/routes/users.js:18]: Should we just require `agency` in the body and not have this fallback that defaults to creating a user within creator's own agency? Client UI could handle setting that as a default value.

File: packages/server/__tests__/api/agencies.test.js
Comment (mattbroussard) [inline packages/server/__tests__/api/agencies.test.js:45]: I think it's worth asserting on what the specific agency IDs in the response are. Same applies to the other tests added in this file.

File: packages/server/__tests__/api/grants.test.js
Comment (mattbroussard) [inline packages/server/__tests__/api/grants.test.js:316]: Is it guaranteed at this point that agency 0 is interested in some grant so we know that undefined here means it was actually filtered out vs. it just doesn't exist?

File: packages/server/__tests__/db/db.test.js
Comment (mattbroussard) [inline packages/server/__tests__/db/db.test.js:32]: Can these tests that look at rows.length check the actual IDs of grants returned?

**Review by mattbroussard — APPROVED** [review]
Mostly minor comments but some questions mainly about:
1. What is the `/api/organizations/:organizationId/agencies/impersonable` endpoint?
2. What are the new semantics for impersonation / is the query in `inTenant` correct?

There may be some merge conflicts with #189, though probably less now that this PR does not include any "super-admin" logic (for which I added some logic over there). Let me know if you'd like to discuss synchronously (ideally today or Monday; I am traveling 9/13-20).

Marking approved to unblock, but please make requested changes before merge & re-request review if things change substantially.

File: packages/server/src/db/index.js
Comment (bspates) [reply packages/server/src/db/index.js:533]: 🙃

File: packages/server/src/db/index.js
Comment (bspates) [reply packages/server/src/db/index.js:810]: There is actually a place that is calling isUserAuthorized outside of the access-helper that was using list of agency ids and couldn't be converted to just using tenant id. https://github.com/usdigitalresponse/usdr-gost/blob/tenant-scoping-apis2/packages/server/src/routes/grants.js#L185 

I'll just make it always an array param, and handle the distinction at the isUserAuthorized level.

File: packages/server/src/routes/users.js
Comment (bspates) [reply packages/server/src/routes/users.js:18]: Since both cases were covered in the existing unit tests it seemed better to follow the pattern than change it and increase the scope of the PR to more front end changes.

File: packages/server/src/db/index.js
Comment (bspates) [reply packages/server/src/db/index.js:567]: This is now changed below this comment.

File: packages/server/__tests__/api/grants.test.js
Comment (bspates) [reply packages/server/__tests__/api/grants.test.js:316]: Yeah it is setup in the db seed run before the tests https://github.com/usdigitalresponse/usdr-gost/blob/tenant-scoping-apis2/packages/server/seeds/dev/ref/grants.js#L119

File: packages/server/src/lib/access-helpers.js
Comment (bspates) [reply packages/server/src/lib/access-helpers.js:26]: Seems like it

File: packages/server/src/lib/access-helpers.js
Comment (bspates) [reply packages/server/src/lib/access-helpers.js]: Good point, I think I missed this nuance after removing the super tenant concept.

File: packages/server/src/routes/agencies.js
Comment (bspates) [reply packages/server/src/routes/agencies.js:24]: So originally I had front-end changes to use this endpoint for the only purpose of selecting another agency to impersonate. I decided to move them to a later PR when taking out the super-tenant stuff and they were a little scrambled together. Should have mentioned that somewhere. I can remove these backend changes from this PR.

File: packages/server/src/lib/access-helpers.js
Comment (mattbroussard) [reply packages/server/src/lib/access-helpers.js]: Huh? So what are the semantics _supposed to be_?

Comment (bspates) [reply packages/server/src/lib/access-helpers.js]: Impersonation is limited to children in your agency tree, which is important when checking the selectedAgency, not just when choosing an agency to impersonate. In almost all other situations users can see objects across all agencies in a tenant. Like the user listing api is not limited by membership in a child agency anymore. So I'll be add a different function that checks for allowable impersonation vs allowing access to a resource... OR ask @mhuang-usdr If its okay if we let admin users (which are the only ones that can impersonate) impersonate any agency in a tenant, to help simplify things.

File: packages/server/src/db/index.js
Comment (mattbroussard) [reply packages/server/src/db/index.js]: Your comment after the updates says this function

> Determines if a user is in an agency and that agency is in the same tenant

However, I'm not sure the query is quite right after playing with it in Postico a bit: it checks that the agency and user both match the specified tenant ID, and but not that the user belongs to that agency. I think you need to add a clause like `.where('users.agency_id', knex.ref('agencies.id'))`. That will ensure you only get rows if the user is part of the one of the specified agencies.

File: packages/server/src/db/index.js
Comment (mattbroussard) [inline packages/server/src/db/index.js:819]: nits on this section:
 - no need to gate on one vs. multiple IDs; a `whereIn` with one item is just fine
 - I don't think you need the querybuilder callback, doesn't `q.whereIn(...)` work?
 - I think empty `agencyIds` ought to throw instead of just return false; it seems like that would always be a programming mistake?

File: packages/server/src/lib/access-helpers.js
Comment (mattbroussard) [reply packages/server/src/lib/access-helpers.js]: Got it. I suspect the answer is we want impersonation to still be subagencies only.

Comment (mhuang-usdr) [reply packages/server/src/lib/access-helpers.js]: Clarifying question on what "allowable impersonation vs allowing access to a resource" means -

Let's say we have this agency tree:

- main agency
   -  A
      - sub-A
   - B

So,  if an admin in Agency A is impersonates sub-A, they should have full write/edit access to sub-A.

**Question:**
- in this current PR, if an admin in Agency A is impersonates Agency B, what is the experience? Is it a read-only experience, or does Agency A have write access?

From the product perspective, read-only impersonation should be fine - what's important is that the user is not able to write to an agency they're not supposed to - so I'd lean towards the simplest engineering solution until we hear future feedback.

However, I'd like to test the user experience of such an impersonation to make sure it's not totally broken!

Comment (mhuang-usdr) [reply packages/server/src/lib/access-helpers.js]: @bspates can I check via the same ngrok link you sent earlier?

File: packages/server/src/db/index.js
Comment (bspates) [reply packages/server/src/db/index.js:121]: It works with any agency ID, which for a recursive tree function it makes sense to refer to any node as a possible "root".

File: packages/server/src/lib/access-helpers.js
Comment (bspates) [reply packages/server/src/lib/access-helpers.js]: Talked with @mhuang-usdr, the answer is that write access in all cases needs to be limited by the agency tree. The nuance here is that with multi-tenancy we have admin users able to see across agency trees in a tenant, BUT not edit them. This PR makes it so that admins can write/edit any object in the tenancy. The work to move to a read but not write across the tenancy is not small, so we're spinning that out to another issue, as this is more of a future use case than a current one.

File: packages/server/src/db/index.js
Comment (bspates) [reply packages/server/src/db/index.js:819]: I didn't realize where and "andWhere" were aliases, so I thought I had to do an an andWhere => whereIn, thanks! An in clause can be a much slower query due to not activating the index or forcing a file sort, sometimes the query engine will optimize it away for a single record sometimes not.

File: packages/server/src/db/index.js
Comment (bspates) [reply packages/server/src/db/index.js]: I think I need to change the comment, its actually whether they have access to said agency, not whether they are apart of it, so its just that both the user and agency are in the same tenant.

File: packages/server/__tests__/db/db.test.js
Comment (bspates) [reply packages/server/__tests__/db/db.test.js:32]: I changed this one to check the date ordering, so the created_at values, since they reference the same grant id. The tests below are actually stats endpoints so they can only verify the count as its all that is returned by the api.

File: packages/server/src/routes/agencies.js
Comment (bspates) [reply packages/server/src/routes/agencies.js:23]: I'd rather not grow the scope of this PR with a bunch of consistency changes that are currently outside of the diff. I'll make another issue for it. The repo in general has a lot of inconsistencies. I've tried to just leave the code in one place consistent with itself.

Comment (bspates) [reply packages/server/src/routes/agencies.js:23]: https://github.com/usdigitalresponse/usdr-gost/issues/328

File: packages/server/seeds/dev/ref/agencies.js
Comment (bspates) [reply packages/server/seeds/dev/ref/agencies.js:14]: Having a _ would mean its not an abbreviation.

File: packages/server/src/lib/access-helpers.js
Comment (bspates) [reply packages/server/src/lib/access-helpers.js:17]: Both are still in use. For the case when an api is trying to do something on behalf of a different user than the one in session.

**Pull request merged**
